onnx_trt_backend.cpp

```diff
@@ -0,0 +1,749 @@
+  int device_id_{0};
+  size_t max_batch_size_{0};
+  size_t batch_size_{0};
+  cudaStream_t stream_;
+};
+
+void GraphRep::ClearDeviceBuffers() {
+  for (auto kv : device_buffers_) {
+    cudaFree(kv.second);
+  }
+  device_buffers_.clear();
+}
+
+onnxStatus GraphRep::CheckAndBindTensor(const nvinfer1::Dims &dims,
+                                        const onnxTensorDescriptor &tensor) {
+  // Check memory type
+  if (tensor.memoryType != ONNXIFI_MEMORY_TYPE_CPU &&
+      tensor.memoryType != ONNXIFI_MEMORY_TYPE_CUDA_BUFFER) {
+  std::cerr << tensor.memoryType << std::endl;
+    return ONNXIFI_STATUS_INVALID_DATATYPE;
+  }
+  // Check input shape
+  auto ret = CheckShape(dims, tensor);
+  if (ret != ONNXIFI_STATUS_SUCCESS) {
+    return ret;
+  }
+
+  // For CPU tensor, we need to create a device memory and the bind. For CUDA
+  // tensor, we can bind directly
+  if (tensor.memoryType == ONNXIFI_MEMORY_TYPE_CPU) {
+    void *cuda_buffer;
+    size_t footprint = GetTensorFootprint(tensor);
+    if (!footprint) {
+      return ONNXIFI_STATUS_INVALID_SHAPE;
+    }
+    if (cudaMalloc(&cuda_buffer, footprint) != cudaSuccess) {
+      return ONNXIFI_STATUS_NO_DEVICE_MEMORY;
+    }
+    device_buffers_.emplace(tensor.name, cuda_buffer);
+    bindings_.push_back(cuda_buffer);
+  } else {
+    bindings_.push_back((void *)(tensor.buffer));
+  }
+
+  return ONNXIFI_STATUS_SUCCESS;
+}
+
+onnxStatus GraphRep::InitIO(uint32_t inputsCount,
+                            const onnxTensorDescriptor *inputDescriptors,
+                            uint32_t outputsCount,
+                            const onnxTensorDescriptor *outputDescriptors) {
+  CudaDeviceGuard guard(device_id_);
+  ClearDeviceBuffers();
+  // Setup the input/output bindings and decide batch size
+  for (unsigned i = 0; i < inputsCount; ++i) {
+    if (!inputDescriptors[i].name) {
+      return ONNXIFI_STATUS_INVALID_NAME;
+    }
+    // We only support NCHW
+    if (inputDescriptors[i].dimensions != 4) {
+      return ONNXIFI_STATUS_INVALID_SHAPE;
+    }
+    if (i == 0) {
+      batch_size_ = inputDescriptors[i].shape[0];
+    } else {
+      if (batch_size_ != inputDescriptors[i].shape[0]) {
+        return ONNXIFI_STATUS_INVALID_SHAPE;
+      }
+    }
+    std::cerr << "Adding input " << i << ": " << inputDescriptors[i].name
+              << ", type: " << inputDescriptors[i].memoryType << std::endl;
+    input_map_.emplace(std::string(inputDescriptors[i].name),
+                       inputDescriptors + i);
+  }
+
+  // We don't support the case when batch size is larger than max batch size
+  // yet, but this is not a hard constraint.
+  if (batch_size_ > max_batch_size_) {
+    return ONNXIFI_STATUS_NO_DEVICE_RESOURCES;
+  }
+
+  for (unsigned i = 0; i < outputsCount; ++i) {
+    if (!outputDescriptors[i].name) {
+      return ONNXIFI_STATUS_INVALID_NAME;
+    }
+    // We only support NCHW
+    if (outputDescriptors[i].dimensions != 4) {
+      return ONNXIFI_STATUS_INVALID_SHAPE;
```

Comment:
This should be `ONNXIFI_STATUS_UNSUPPORTED_SHAPE`. `ONNXIFI_STATUS_UNSUPPORTED_*` codes are returned when an argument is reasonable, but not supported by this particular backend. `ONNXIFI_STATUS_INVALID_*` codes are returned when an argument makes no sense (e.g. undefined enum value, 0 in shape dimensions, corrupted protobuf format, etc)